src/Processors/QueryPlan/Optimizations/optimizeTree.cpp

```diff
@@ -112,6 +112,9 @@ void optimizeTreeSecondPass(const QueryPlanOptimizationSettings & optimization_s
 
         if (frame.next_child == 0)
         {
+            /// insert sorting for join if needed before optimizeReadInOrder to optimize them
```

Comment:
It's inside `applyOrderForJoin`. If optimization is enabled we will try to applly it, otherwise insert sorting steps. Previously it was inserted unconditionally on plan building stage.